dialogue-core/src/main/java/com/palantir/dialogue/core/BalancedNodeSelectionStrategyChannel.java

```diff
@@ -92,25 +116,101 @@
 
         // TODO(dfox): P2C optimization when we have high number of nodes to save CPU?
         // http://www.eecs.harvard.edu/~michaelm/NEWWORK/postscripts/twosurvey.pdf
-        SortableChannel[] sortedChannels = sortByScore(preShuffled);
+        ChannelStats[] sortableChannels = computeScores(preShuffled);
+        Arrays.sort(sortableChannels, BY_SCORE);
 
-        for (SortableChannel channel : sortedChannels) {
+        for (ChannelStats channel : sortableChannels) {
             Optional<ListenableFuture<Response>> maybe = channel.delegate.maybeExecute(endpoint, request);
             if (maybe.isPresent()) {
+                maybeSampleAllChannelRttsNonBlocking();
                 return maybe;
             }
         }
 
         return Optional.empty();
     }
 
-    private static SortableChannel[] sortByScore(List<MutableChannelWithStats> preShuffled) {
-        SortableChannel[] sorted = new SortableChannel[preShuffled.size()];
+    private static ChannelStats[] computeScores(List<MutableChannelWithStats> preShuffled) {
+        ChannelStats[] snapshotArray = new ChannelStats[preShuffled.size()];
+
+        long bestRttNanos = Long.MAX_VALUE;
```

Comment:
I'd decouple the logic of constructing the `snapshotArray` from computing the `rttSpecturom` and factor this out so we have a single function that returns to you an OptionalLong which is the `rttSpectrum`